### PR TITLE
Dev-oriented scripts to build and deploy the RH distribution to minishift, based on CICO script

### PR DESCRIPTION
**Pull request by davidfestal**
Developer-oriented scripts that allow building / deploying / rolling out the RH distribution to a local minishift installation:

- `dev_build.sh`:
    - run the RH maven build with the options passed as arguments,
    - generate the docker images for the upstream and the RH
distribution and tag them appropriately in the *local docker
environment*.

- `dev_build_and_deploy.sh`:
    - run the RH maven build with the options passed as arguments,
    - generate the docker images for the upstream and the RH
distribution and tag them appropriately in the *minishift docker
environment*
    - delete all the Che-related resources in minishift
    - create all the require resources in minishift based on the
generated RH docker image.
    
- `dev_build_and_rollupdate.sh`:
    - run the RH maven build with the options passed as arguments,
    - generate the docker images for the upstream and the RH
distribution and tag them appropriately in the *minishift docker
environment*
    - roll-update the installed Che in minishift with the generated RH
docker image.

### CODE REVIEW

File: config [hunk 1]
```diff
@@ -27,3 +25,4 @@ export RH_CHE_DOCKER_IMAGE_VERSION=${RH_CHE_DOCKER_IMAGE_VERSION:-${DEFAULT_RH_C
 export DOCKER_HUB_USER=${DOCKER_HUB_USER:-${RHCHEBOT_DOCKER_HUB_USER}}
 export DOCKER_HUB_PASSWORD=${DOCKER_HUB_PASSWORD:-${RHCHEBOT_DOCKER_HUB_PASSWORD}}
 export DOCKER_HUB_NAMESPACE=${DOCKER_HUB_NAMESPACE:-${RH_CHE_DOCKER_HUB_NAMESPACE}}
+export DOCKER_HUB_REGISTRY_PREFIX=${DOCKER_HUB_REGISTRY_PREFIX:-""}
```
Comment (l0rd) [inline config]: Is it really necessary to introduce that new env variable. How are you using it? Can you provide an example?

Comment (davidfestal) [inline config]: I have a RHEL CSB laptop, and the CSB docker setup has the RH registry as primary docker registry. So by default when tagging or pushing without the `docker.io/` registry prefix, these actions are done against the `registry.access.redhat.com` registry, and push fails with an authentication failure. In such a use-case, being able to explicitly set the registry when tagging / pulling in the scripts helps a lot.
But of course the default value fro this is empty, so that it doesn't change anything for people that don't suffer of this RHEL CSB docker preset.

**Comment by ibuziuk** [comment]
@davidfestal is it the correct way for running the script:

`./dev_build.sh -DwithoutKeycloak -Pfast -DlocalCheRepository=${UPSTREAM_CHE_PATH}`

**Comment by ibuziuk** [comment]
after running the command above build failed with the following error:
```

[INFO] [INFO] Fabric8 IDE :: Compiling GWT Application ........... FAILURE [02:10 min]
[INFO] [INFO] Fabric8 IDE :: Compiling WS Master WAR ............. SKIPPED
[INFO] [INFO] Fabric8 IDE :: Che Dashboard WAR ................... SKIPPED
[INFO] [INFO] Fabric8 IDE :: Assemblies Tomcat ................... SKIPPED
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] BUILD FAILURE
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Total time: 02:55 min
[INFO] [INFO] Finished at: 2017-05-24T16:15:23+02:00
[INFO] [INFO] Final Memory: 91M/786M
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [ERROR] Failed to execute goal org.codehaus.mojo:gwt-maven-plugin:2.8.0:compile (default) on project fabric8-ide-assembly-ide-war: Command [[
[INFO] [ERROR] /bin/sh -c '/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java' '-Xss512m' '-Xmx2048m' 'com.google.gwt.dev.Compiler' '-logLevel' 'INFO' '-war' '/home/ilya/git/jbosstools-git/rh-che/target/builds/fabric8/fabric8-che/assembly/assembly-ide-war/target/fabric8-ide-assembly-ide-war-5.6.0-openshift-connector-fabric8-SNAPSHOT' '-localWorkers' '2' '-XfragmentCount' '-1' '-sourceLevel' '1.8' '-gen' '/home/ilya/git/jbosstools-git/rh-che/target/builds/fabric8/fabric8-che/assembly/assembly-ide-war/target/.generated' 'org.eclipse.che.ide.IDE'
```

**Comment by davidfestal** [comment]
> @davidfestal is it the correct way for running the script:
>
> `./dev_build.sh -DwithoutKeycloak -Pfast -DlocalCheRepository=${UPSTREAM_CHE_PATH}`

well, prefer `-Dfast` instead of `-Pfast` when passing your arguments to the scripts (since profile activation cannot be passed as several arguments).

You can also add the `clean` at the end if necessary. The script only adds the `install` at the end of the maven comand.

**Comment by davidfestal** [comment]
@ibuziuk 
> after running the command above build failed with the following error:

could you share somewhere more complete logs: this would help understanding why GWT is failing ?

File: config (hunk 1 above)
Comment (l0rd) [inline config]: Can't you just set `DOCKER_HUB_NAMESPACE` in this case? 
For example `export DOCKER_HUB_NAMESPACE=docker.io/rhchestage`

**Comment by ibuziuk** [comment]
@davidfestal the failure was probably coupled with my plugin. After building from master the image was build \o/

**Comment by davidfestal** [comment]
> After building from master the image was build \o/

@ibuziuk good news !

**Comment by ibuziuk** [comment]
@davidfestal hmmm.... the script produces a bunch of images:

```
eclipse/che-server                  nightly                                a1d3d8e3f42d        7 minutes ago       773MB
rhchestage/che-server               0ae26d7-fabric8-no-dashboard-c3b7383   a1d3d8e3f42d        7 minutes ago       773MB
rhchestage/che-server               nightly-fabric8-no-dashboard           a1d3d8e3f42d        7 minutes ago       773MB
rhchestage/che-server               0ae26d7-fabric8-c3b7383                90d6a274490f        8 minutes ago       780MB
rhchestage/che-server               nightly-fabric8                        90d6a274490f        8 minutes ago       780MB
rhchestage/che-server               0ae26d7                                b39fd101e1ad        8 minutes ago       789MB
rhchestage/che-server               nightly                                b39fd101e1ad        8 minutes ago       789MB
```
which one should be used for deployment on minishift ?

File: config (hunk 1 above)
Comment (davidfestal) [inline config]: I probably could have done that, yes. I should test this locally to confirm though

**Comment by davidfestal** [comment]
> which one should be used for deployment on minishift ?

nightly-fabric8

But if you use the `dev_build_and_deploy.sh` with minishift running on your host, it will deploy it automatically (more precisely delete all and recreate all  the Che environment with your newly built image)

File: config (hunk 1 above)
Comment (l0rd) [inline config]: Ok please test that. Adding a new variable makes the script more complicated. What you can do too is setting `export RH_CHE_DOCKER_HUB_NAMESPACE=docker.io/rhchestage` by default. That should work on both RHEL and non RHEL distributions.

**Comment by l0rd** [comment]
I would move
 
```
build_che.shdev_build.sh
dev_build_and_deploy.sh
dev_build_and_rollupdate.sh
```

to the `scripts` folder. And delete `scripts/build.sh` that is not useful anymore.

And what is `setup-env-for-local-deployment-test.sh` for? Can we remove that too?

**Comment by ibuziuk** [comment]
`./dev_build_and_deploy.sh -DwithoutKeycloak -DwithoutDashboard -Dfast -DlocalCheRepository=${UPSTREAM_CHE_PATH} clean`

@davidfestal is it expected that the following command deploys che with dashboard on minishift ?

**Comment by ibuziuk** [comment]
@davidfestal oh, it might be because my build failed during the `dev_build_and_deploy.sh` and previous image has been deployed:
```
[INFO] [INFO] # Native memory allocation (mmap) failed to map 188219392 bytes for committing reserved memory.
[INFO] [INFO] # An error report file with more information is saved as:
[INFO] [INFO] # /home/ilya/git/jbosstools-git/rh-che/target/builds/fabric8-without-dashboard/fabric8-che/hs_err_pid30618.log
[INFO] [ERROR] OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x000000009c880000, 188219392, 0) failed; error='Cannot allocate memory' (errno=12)
```
Even though the build failed `dev_build_and_deploy.sh` continue execution - should on it be terminated after mvn build failure ?

**Comment by davidfestal** [comment]
It always deploys the one with dashboard. It's a limitation but I'll fix that asap.

**Comment by davidfestal** [comment]
I'll look into the case of the failed build also.

**Comment by davidfestal** [comment]
> I would move
>  
> ```
> build_che.shdev_build.sh
> dev_build_and_deploy.sh
> dev_build_and_rollupdate.sh
> ```
> 
> to the `scripts` folder. And delete `scripts/build.sh` that is not useful anymore.
>

OK. I'll do it.

> 
> And what is `setup-env-for-local-deployment-test.sh` for? Can we remove that too?

it was to test `cico-deploy.sh` locally in a row CentOS vm. It could be removed, or renamed and moved into the `scripts` folder if still useful in some rare cases.

File: config (hunk 1 above)
Comment (davidfestal) [inline config]: well, there's also the following line:
```
    docker login -u ${DOCKER_HUB_USER} -p $DOCKER_HUB_PASSWORD -e noreply@redhat.com  ${DOCKER_HUB_REGISTRY_PREFIX}
```
and for this line the solution you propose won't work afaict.

However I assume this line should be moved to the condition executed when it's a CI build (non-developer build).
In the developer use-cases (`dev_buildXXX` scripts), we don't push docker images to any remote repository, so we don't need to login afaik.

Comment (davidfestal) [inline config]: I'll test that

**Comment by ibuziuk** [comment]
> And delete scripts/build.sh that is not useful anymore.


@l0rd what about building `openshift-connector` without red hat assembly ? I think build.sh would be useful for it

**Comment by davidfestal** [comment]
@ibuziuk 

> It always deploys the one with dashboard. It's a limitation but I'll fix that asap.

Fixed by commit cef3968980ea0b6a306073a2d2e4da12588811e4

**Comment by davidfestal** [comment]
@ibuziuk 

> Even though the build failed dev_build_and_deploy.sh continue execution - should on it be terminated after mvn build failure ?

Should be fixed by commit 1bd3291577c7c92666ee67547b7018c39f24fd09

File: config (hunk 1 above)
Comment (davidfestal) [inline config]: @l0rd implemented in commit 66af71d505ab81c314b23088bfa000a2b734328d

**Comment by davidfestal** [comment]
@l0rd what about `run_che.sh` does it still have sense (comes from the initial CI build repo afaik)

**Comment by davidfestal** [comment]
@ibuziuk I rebased the branch on master.

**Comment by l0rd** [comment]
> What about renaming this directory dev-scripts to make it clear these are utility scripts for the developer ?
@davidfestal that's fine but then you should remove the dev prefix for scripts

**Comment by davidfestal** [comment]
> @davidfestal that's fine but then you should remove the dev prefix for scripts

sure that was the idea :-)

**Comment by l0rd** [comment]
> @l0rd what about building openshift-connector without red hat assembly ? I think build.sh would be useful for it

@ibuziuk ok we can rename it `build-openshift-connector.sh` or just integrate that in the `build.sh`?

**Comment by l0rd** [comment]
> >And what is setup-env-for-local-deployment-test.sh for? Can we remove that too?

> it was to test cico-deploy.sh locally in a row CentOS vm. It could be removed, or renamed and moved into the scripts folder if still useful in some rare cases.

@davidfestal ok let's rename it and move it to folder `scripts`

**Comment by davidfestal** [comment]
> @davidfestal ok let's rename it and move it to folder scripts

@l0rd oh sorry, in the meantime I removed it, thinking it was so useful finally :-/

**Comment by l0rd** [comment]
> @l0rd oh sorry, in the meantime I removed it, thinking it was so useful finally :-/

No problem if you have removed it. That's ok

**Comment by davidfestal** [comment]
> > @l0rd what about building openshift-connector without red hat assembly ? I think build.sh would be useful for it
> 
> @ibuziuk ok we can rename it `build-openshift-connector.sh` or just integrate that in the `build.sh`?  

 When you modify the `openshift-connector` code in upstream, don't you want to *also* rebuild the RH distribution for the whole to be consistent ?

IMO the new `build.sh` script is just  a matter of running:
- the right maven modules in the upstream project
- the right dev_XXX script with pre-defined arguments

I'll try to provide a script right now, so that you can tell it that's OK for you

**Comment by ibuziuk** [comment]
`When you modify the openshift-connector code in upstream, don't you want to also rebuild the RH distribution for the whole to be consistent ?`
Not necessarily. When I work on `openshift-connector` in most cases I do not need red hat assembly at all

**Comment by ibuziuk** [comment]
`./dev_build_and_deploy.sh -DwithoutKeycloak -DwithoutDashboard -Dfast -DlocalCheRepository=${UPSTREAM_CHE_PATH} clean`

@davidfestal this command seems to still deploy che-server with dashboard for me

**Comment by davidfestal** [comment]
> `./dev_build_and_deploy.sh -DwithoutKeycloak -DwithoutDashboard -Dfast -DlocalCheRepository=${UPSTREAM_CHE_PATH} clean`
> 
> @davidfestal this command seems to still deploy che-server with dashboard for me

Strange: I had tested it locally. But I'll look into it as soon as I've finished my my current change.

**Comment by ibuziuk** [comment]
@davidfestal this might be because I had the following env vars set ?
```
export CHE_IMAGE_REPO=rhche/che-server
export CHE_IMAGE_TAG=nightly
```

**Comment by davidfestal** [comment]
@l0rd @ibuziuk is it OK for you ? as soon as it is I'll update the Readme.

**Comment by ibuziuk** [comment]
@davidfestal after more testing:
`./build_fabric8_and_deploy.sh  -DwithoutDashboard -Dfast clean`
After running the following command on fresh minishift cluster image che-server image deployment does not work for me:

image

I think there is a bug in the script and invalid image is used for deployment

**Comment by davidfestal** [comment]
> I think there is a bug in the script and invalid image is used for deployment

@ibuziuk this should be fixed now.

File: cico_do_docker_tag.sh [hunk 2]
```diff
@@ -0,0 +1,89 @@
+#!/bin/bash
```
Comment (l0rd) [inline cico_do_docker_tag.sh]: This file should be renamed in `cico_do_docker_build_and_push.sh` as these are the more important operations.

**Comment by davidfestal** [comment]
@l0rd I renamed the *docker-build-related* scripts.

**Comment by ibuziuk** [comment]
@davidfestal not all renaming is done
`./build_fabric8_and_deploy.sh -DwithoutKeycloak -DwithoutDashboard -Dfast clean
`

> bash: cico_do_docker_tag.sh: No such file or directory
> Build Failed!
> Build Failed!

**Comment by davidfestal** [comment]
@ibuziuk oh sorry

**Comment by davidfestal** [comment]
@ibuziuk should be better this time

**Comment by ibuziuk** [comment]
:+1:

**Comment by davidfestal** [comment]
@l0rd is it OK for you to merge ?

File: cico_do_docker_build_tag_push.sh [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    
+    # We are not pushing the nightly tag because we don't need it and CI has an issue
+    # when publishing > 1 tag at a time 
+    # docker push ${DOCKER_HUB_NAMESPACE}/che-server:${NIGHTLY}
+    echo 'export CHE_SERVER_DOCKER_IMAGE_TAG='${TAG} >> ~/che_image_tag.env
+    docker push ${DOCKER_HUB_NAMESPACE}/che-server:${TAG}
+    
+    if [ "${DOCKER_HUB_USER}" == "${RHCHEBOT_DOCKER_HUB_USER}" ]; then
+    # lets also push it to registry.devshift.net
+      docker tag ${DOCKER_HUB_NAMESPACE}/che-server:${NIGHTLY} registry.devshift.net/che/che:${NIGHTLY}
```
Comment (mlabuda) [inline cico_do_docker_build_tag_push.sh]: should we also push stg images to registry devshift? Maybe this should be done later, when promoting an image, right after verification job.

Comment (davidfestal) [inline cico_do_docker_build_tag_push.sh]: @mlabuda would it be possible to do this as a separate issue, after this PR is merged ?

Comment (mlabuda) [inline cico_do_docker_build_tag_push.sh]: Ok, so far it was done this, not right, way so we can do it in separate PR.

Comment (davidfestal) [inline cico_do_docker_build_tag_push.sh]: could you approve the request then ?

Comment (mlabuda) [inline cico_do_docker_build_tag_push.sh]: Done

Comment (davidfestal) [inline cico_do_docker_build_tag_push.sh]: Thanks !

**Comment by davidfestal** [comment]
@l0rd the readme is updated.

**Comment by davidfestal** [comment]
@l0rd any news about the ability to merge ?

File: README.md [hunk 4]
```diff
@@ -256,41 +269,110 @@ scripts/build.sh
 
 `git clone https://github.com/fabric8io/fabric8-online.git`
 
-* Set some environment variables (or run `. ./scripts/setenv-for-deploy.sh`)
+* You may need to override the full path to the `fabric8-online` binary (provided by the `FABRIC8_ONLINE_PATH`
+environment variable), according to where you cloned the `fabric8-online` git repository :
+
```
Comment (l0rd) [inline README.md]: I would remove "may". And that is not a full path to a binary but a path to a repository where yaml for Che deployment are saved:
`s/You may need to override the full path to the fabric8-online binary/You need to set the local path to the fabric8-online git repo/`

File: README.md [hunk 5]
```diff
@@ -224,29 +242,24 @@ cd che
...
+export UPSTREAM_CHE_PATH="/Users/mariolet/Github/che"
 ```
 
-* CHE_IMAGE_REPO and CHE_IMAGE_TAG are the Docker image name and tag which will be built.
-* GITHUB_REPO points to the local directory where the official Che repo was checked out
-
-### Build Che
-
-In the rh-che repository, build Che by executing the following script:
+    * `DOCKER_HUB_NAMESPACE`, `DOCKER_HUB_USER` and `DOCKER_HUB_PASSWORD` can be overridden to point
```
Comment (l0rd) [inline README.md]: This bullet point is not parsed well by the markdown interpreter. It's not at the same level of `UPSTREAM_CHE_PATH` and it looks like `code`

File: README.md [hunk 6]
```diff
@@ -256,41 +269,110 @@ scripts/build.sh
...
+    - runs the RH maven build with the options passed as arguments,
+    - generates the docker images for the upstream and the RH
+distribution and tag them appropriately in the *local docker
+environment*.
+
+- `dev-scripts/build_fabric8_and_deploy.sh`:
+    - runs the RH maven build with the options passed as arguments,
+    - generates the docker images for the upstream and the RH
+distribution and tag them appropriately in the *minishift docker
+environment*
```
Comment (l0rd) [inline README.md]: you can summarize the first two points with just
- runs `dev-scripts/build_fabric8.sh`

I would not mention "in the local (or minishift) docker environment" because you really don't know if the user has previously run `eval $(minishift docker-env)`.

File: README.md [hunk 7]
```diff
@@ -256,41 +269,110 @@ scripts/build.sh
...
+    - deletes all the Che-related resources in minishift
+    - creates all the require resources in minishift based on the
+generated RH docker image.
+    
+- `dev-scripts/dev_build_and_rollupdate.sh`:
+    - runs the RH maven build with the options passed as arguments,
+    - generates the docker images for the upstream and the RH
+distribution and tag them appropriately in the *minishift docker
```
Comment (l0rd) [inline README.md]: Same comment as above. And you have probably forgotten a line.

File: README.md [hunk 8]
```diff
@@ -256,41 +269,110 @@ scripts/build.sh
...
+    - creates all the require resources in minishift based on the
+generated RH docker image.
+    
+- `dev-scripts/build_openshift_connector_and_rollupdate.sh`:
+    - runs the upstream Che maven build on the openshift connector
+    related modules
+    - updates the RH distribution by running RH maven build with
+    the options passed as arguments,
+    - generates the docker images for the upstream and the RH
+distribution and tag them appropriately in the *minishift docker
```
Comment (l0rd) [inline README.md]: same comment as above

File: README.md [hunk 9]
```diff
@@ -224,29 +242,24 @@ cd che
...
 ```
 
-* Set some environment variables (or run `. ./scripts/setenv-for-build.sh`)
+* Set some environment variables:
+    * `UPSTREAM_CHE_PATH` points to the local directory where the official Che repo was checked out
+    and is *mandatory* for some scripts
 
 ```bash
-export CHE_IMAGE_REPO=rhche/che-server
-export CHE_IMAGE_TAG=nightly
```
Comment (l0rd) [inline README.md]: Why are you removing these? How a user is going set the name and tag of the Docker image?

Comment (davidfestal) [inline README.md]: The name and tag of the docker images we want to push to Minishift is now automatically set in the build and deploy scripts according to:
- the variable `DOCKER_HUB_NAMESPACE`
- the fact that the we build *with or without* the dashboard (thanks to the `-DwithoutDashboard` build argument)

The code can be see here for example: https://github.com/davidfestal/rh-che/blob/f3c7bfbafcaacd9f8e8920b578ed1a00af166582/dev-scripts/build_fabric8_and_deploy.sh#L18

File: README.md (hunk 6 above)
Comment (davidfestal) [inline README.md]: `eval $(minishift docker-env)` is already called in the scripts that deploy the docker image to minishift (see [here](https://github.com/davidfestal/rh-che/blob/f3c7bfbafcaacd9f8e8920b578ed1a00af166582/dev-scripts/build_fabric8_and_deploy.sh#L10))

This means that the images created and tagged by the `build_xxx_and_deploy.sh` (or `build_xxx_and_rollupdate`) scripts are created and tagged *in the minishift docker environment*.

This is not the case if you run `build_fabric8.sh` (or `build_openshift_connector.sh`) directly: the docker image and tags are created in your *local* docker environment because `eval $(minishift docker-env)` is not called in this case.

Let me add a point:
- Previously `eval $(minishift docker-env)` was called as part of `setenv-for-build.sh`. But sourcing this file explicitely before running `build.sh` was changing the docker environment of the current shell session without notifying. The only way to come back to the local docker environment was to manually unset some variables. This can lead to strange things and problems until noticed (at least this was the case for me).
- Now  `eval $(minishift docker-env)` is only called by scripts that are not "sourced" but executed. So the docker environment is only changed *during the script execution*, without impacting the current shell and further docker calls.

File: README.md (hunk 9 above)
Comment (l0rd) [inline README.md]: Ok sorry you are right

**Comment by davidfestal** [comment]
Thanks you for your proof-reading  @mario :-)
